### PR TITLE
[MockProver] Add row and col count to cost model

### LINKED ISSUE #638: Method to print no. of rows used by layouter
Currently, there is no easy way to access the number of rows used by a layouter (short of printing it out). Calculating the number of rows would require doing at least a measurement pass of the layouter.

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Adds detailed row and column count to cost model to aid iteration and development of floor planners.

Closes #638. Also cleans up codebase slightly by merging the Assembly/Layout abstraction used by the graphical renderer vs. the cost model

Example cost model measurement now:
```javascript
{
  k: 9,
  max_deg: 29,
  advice_columns: 11,
  instance_queries: 3,
  advice_queries: 31,
  fixed_queries: 5,
  lookups: 1,
  permutation_cols: 8,
  point_sets: 4,
  max_rows: 48,
  max_advice_rows: 48,
  max_fixed_rows: 11,
  num_fixed_columns: 5,
  num_advice_columns: 11,
  num_instance_columns: 3,
  num_total_columns: 19,
  _marker: PhantomData,
}
```

### CODE REVIEW

File: halo2_proofs/src/dev/cost.rs [hunk 1]
```diff
@@ -38,34 +39,151 @@ pub struct CircuitCost<G: PrimeGroup, ConcreteCircuit: Circuit<G::Scalar>> {
...
+    pub fn new(k: u32, n: usize, num_selectors: usize) -> Self {
+        Layout {
+            k,
+            regions: vec![],
+            current_region: None,
+            total_rows: 0,
+            total_advice_rows: 0,
+            total_fixed_rows: 0,
+            /// Any cells assigned outside of a region. We store this as a `Vec` so that if any
+            /// cells are double-assigned, they will be visibly darker.
```
Comment (daira) [inline halo2_proofs/src/dev/cost.rs:106]: These comments don't need to be duplicated here.

File: halo2_proofs/src/dev/cost.rs [hunk 2]
```diff
@@ -135,21 +262,22 @@ impl<G: PrimeGroup, ConcreteCircuit: Circuit<G::Scalar>> CircuitCost<G, Concrete
...
+    pub fn measure(k: u32, circuit: &ConcreteCircuit) -> Self {
         // Collect the layout details.
         let mut cs = ConstraintSystem::default();
         let config = ConcreteCircuit::configure(&mut cs);
-        let mut assembly = Assembly {
-            selectors: vec![vec![false; 1 << k]; cs.num_selectors],
-        };
+        // let mut assembly = Assembly {
+        //     selectors: vec![vec![false; 1 << k]; cs.num_selectors],
+        // };
```
Comment (daira) [inline halo2_proofs/src/dev/cost.rs:271]: Just delete this.

**Review by daira — APPROVED** [review]
utACK with non-blocking suggestions.

File: halo2_proofs/src/dev/cost.rs (hunk 1 above)
Comment (nalinbhardwaj) [reply halo2_proofs/src/dev/cost.rs]: Fixed!

File: halo2_proofs/src/dev/cost.rs (hunk 2 above)
Comment (nalinbhardwaj) [reply halo2_proofs/src/dev/cost.rs]: Deleted

**Review by daira — APPROVED** [review]
re-utACK

**Comment by daira** [comment]
This needs a rustfmt.

**Pull request merged**